File: backtracking/word_search.js
/**
 * @param {character[][]} board
 * @param {string} word
 * @returns {boolean}
 */
var exist = function(board, word) {
    const valid = (row, col) => {
        return 0 <= row && row < nrows && 0 <= col && col < ncols;
    }

    const backtrack = (row, col, i, seen) => {
        if (i == word.length)
        {
            return true;
        }

        for (const [dcol, drow] of directions)
        {
            const nextRow = row + drow;
            const nextCol = col + dcol;

            if (valid(nextRow, nextCol) && !seen[nextRow][nextCol])
            {
                if (board[nextRow][nextCol] == word[i])
                {
                    seen[nextRow][nextCol] = true;

                    if (backtrack(nextRow, nextCol, i + 1, seen))
                    {
                        return true;
                    }

                    seen[nextRow][nextCol] = false;
                }
            }
        }

        return false;
    }

    const directions = [[0, 1], [1, 0], [0, -1], [-1, 0]];
    const nrows = board.length;
    const ncols = board[0].length;

    for (let row = 0; row < nrows; row++)
    {
        for (let col = 0; col < ncols; col++)
        {
            let seen = [];
            for (let i = 0; i < nrows; i++)
            {
                seen.push(new Array(ncols).fill(false));
            }

            seen[row][col] = true;
            if (board[row][col] == word[0] && backtrack(row, col, 1, seen))
            {
                return true;
            }
        }
    }

    return false;
};

File: backtracking/word_search.py
def exist(board: list[list[str]], word: str) -> bool:
    # Space complexity is O(L) where L = len(word) for seen set
    # and recursion call stack

    def valid(row: int, col: int) -> bool:
        return 0 <= row < nrows and 0 <= col < ncols

    def backtrack(row: int, col: int, i: int, seen: list[list[bool]]) -> bool:
        if i == len(word):
            return True

        for dcol, drow in directions:
            next_row, next_col = row + drow, col + dcol

            if valid(next_row, next_col) and (next_row, next_col) not in seen:
                if board[next_row][next_col] == word[i]:
                    seen.add((next_row, next_col))

                    if backtrack(next_row, next_col, i + 1, seen):
                        return True

                    seen.remove((next_row, next_col))

        return False

    directions = [(0, 1), (1, 0), (0, -1), (-1, 0)]
    nrows = len(board)
    ncols = len(board[0])

    for row in range(nrows):
        for col in range(ncols):
            if (
                board[row][col] == word[0]
                and backtrack(row, col, 1, {(row, col)})
            ):
                return True

    return False
